modules/src/ics04_channel/context.rs

```diff
@@ -0,0 +1,130 @@
+//! ICS4 (channel) context. The two traits `ConnectionReader` and `ConnectionKeeper` define
+//! the interface that any host chain must implement to be able to process any `ChannelMsg`.
+//! TODO make "ADR 004: IBC protocol implementation" for more details.
+//!
+use crate::ics02_client::client_def::{AnyClientState, AnyConsensusState};
+use crate::ics03_connection::connection::ConnectionEnd;
+use crate::ics04_channel::channel::{ChannelEnd, State};
+use crate::ics04_channel::error::Error;
+use crate::ics04_channel::handler::ChannelResult;
+use crate::ics04_channel::version::{get_compatible_versions, pick_version};
+use crate::ics05_port::capabilities::Capability;
+use crate::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::Height;
+
+/// A context supplying all the necessary read-only dependencies for processing any `ChannelMsg`.
+pub trait ChannelReader {
+    /// Returns the ChannelEnd for the given identifier `chan_id`.
+    fn channel_end(&self, port_channel_id: &(PortId, ChannelId)) -> Option<ChannelEnd>;
+
+    /// Returns the ConnectionState for the given identifier `connection_id`.
+    fn connection_state(&self, connection_id: &ConnectionId) -> Option<ConnectionEnd>;
+
+    fn connection_channels(&self, cid: &ConnectionId) -> Option<Vec<(PortId, ChannelId)>>;
+
+    fn channel_client_state(&self, port_channel_id: &(PortId, ChannelId))
+        -> Option<AnyClientState>;
+
+    fn channel_consensus_state(
```

Comment:
I don't see this function used anywhere. I'm wondering what's the use for it.

Reply:
I looked into https://github.com/informalsystems/ibc-rs/blob/master/proto/src/prost/ibc.core.channel.v1.rs
and there at line 497 there's a QueryChannelConsensusStateResponse so I put it (tentatively) in the interface for the query implementation.

Reply:
Ah, cool.

As a general comment, it's usually a good idea to avoid making a PR too big, otherwise it is difficult to review and understand it. So it's preferable to only bring in changes which are necessary.

For this method specifically, I don't understand why would it need to be part of the `ChannelReader` trait. The trait captures all methods necessary for the [channel handlers](https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics#sub-protocols), and I'm not convinced this method is needed there (for any handler). What do you think?

Reply:
> IMO this is the right place for the implementation

Hmm, Reader/Keeper captures dependencies needed to execute/handle transactions. These are entirely different from handling queries. If the IBC module will be called for handling queries, then it may make sense to involve the Keeper, and we can keep queries code here, but this was not clear to me.